src/security/collection.cpp

```diff
@@ -0,0 +1,396 @@
+        indexed_entries;
+    for (it.reset(); !it.ended(); ++it) {
+      auto [key, value] = *it;
+      std::size_t const num_occurr = keys_bag[key];
+
+      // common scenario: only 1 occurrence of the key
+      if (num_occurr == 1) {
+        dnsec::ddwaf_obj &entry = *next_free_entry++;
+        entry.set_key(key);
+        entry.make_string(value);
+        continue;
+      }
+
+      auto ie = indexed_entries.find(key);
+      if (ie == indexed_entries.end()) {
+        // first occurrence of this key
+        dnsec::ddwaf_obj &entry = *next_free_entry++;
+
+        entry.set_key(key);
+        auto &arr_val = entry.make_array(num_occurr, memres_);
+        indexed_entries.insert({key, &arr_val});
+
+        if (!it.is_delete()) {
+          arr_val.at_unchecked<dnsec::ddwaf_obj>(0).make_string(value);
+          entry.nbEntries = 1;
+        }
+      } else {
+        // subsequent occurrence of this key
+        auto &arr_val = *ie->second;
+        if (!it.is_delete()) {
+          arr_val.template at_unchecked<dnsec::ddwaf_obj>(arr_val.nbEntries++)
+              .make_string(value);
+        } else {
+          arr_val.nbEntries = 0;
+        }
+      }
+    }
+  }
+
+  static void set_request_uri_raw(const ngx_http_request_t &request,
+                                  dnsec::ddwaf_obj &slot) {
+    set_map_entry_str(slot, kUriRaw, request.unparsed_uri);
+  }
+
+  static void set_request_method(const ngx_http_request_t &request,
+                                 dnsec::ddwaf_obj &slot) {
+    set_map_entry_str(slot, kMethod, request.method_name);
+  }
+
+  // adapt to the same iteratror format as query_string_iter
+  template <bool IsRequest>
+  struct HeaderKeyValueIter {
+    HeaderKeyValueIter(const ngx_list_t &list, std::string_view exclude,
+                       dnsec::DdwafMemres &memres)
+        : list_{list},
+          memres_{memres},
+          exclude_{exclude},
+          it_{list},
+          end_{dnsec::NginxListIter<ngx_table_elt_t>::end(list)} {}
+
+    void reset() { it_ = decltype(it_){list_}; }
+
+    bool ended() const noexcept { return !(it_ != end_); }
+
+    HeaderKeyValueIter &operator++() {
+      while (true) {
+        ++it_;
+        if (!(it_ != end_)) {
+          break;
+        }
+        const auto &h = *it_;
+        auto lc_key = safe_lowcase_key(h);
+        if (lc_key != exclude_) {
+          break;
+        }
+        // then it's the excluded key; continue
+      }
+      return *this;
+    }
+
+    std::string_view cur_key() {
+      const auto &h = *it_;
+      return safe_lowcase_key(h);
+    }
+
+    std::pair<std::string_view, std::string_view> operator*() {
+      const auto &h = *it_;
+      return {safe_lowcase_key(h), to_string_view(h.value)};
+    }
+
+    bool is_delete() const {
+      if constexpr (IsRequest) {
+        return false;
+      } else {  // response headers
+        const auto &h = *it_;
+        return h.hash == 0;
```

Comment:
Still puzzled of what this means... can't the iterator skip this entry during `operator++`?

Reply:
It can't be skipped because it signals important information (that the header previously defined should be deleted). It _could_ be skipped only if all the references to the header up to the _last_ deletion would be deleted as well. But that would require a first pass and saving this information somewhere.